providers/siw/siw.c

```diff
@@ -0,0 +1,941 @@
+		goto fail;
+	}
+	if (attr->srq) {
+		qp->srq = srq_base2siw(attr->srq);
+	} else if (resp.siw.rq_key != SIW_INVAL_UOBJ_KEY) {
+		int rq_size = resp.siw.num_rqe * sizeof(struct siw_rqe);
+
+		qp->recvq = mmap(NULL, rq_size, PROT_READ | PROT_WRITE,
+				 MAP_SHARED, base_ctx->cmd_fd, resp.siw.rq_key);
+
+		if (qp->recvq == MAP_FAILED) {
+			if (siw_debug)
+				printf("libsiw: RQ mapping failed: %d\n",
+				       resp.siw.num_rqe);
+			qp->recvq = NULL;
+			goto fail;
+		}
+	} else {
+		if (siw_debug)
+			printf("libsiw: prepare RQ mapping failed\n");
+		goto fail;
+	}
+	qp->db_req.qp_handle = qp->base_qp.handle;
+
+	return &qp->base_qp;
+fail:
+	ibv_cmd_destroy_qp(&qp->base_qp);
+
+	if (qp->sendq)
+		munmap(qp->sendq, qp->num_sqe * sizeof(struct siw_sqe));
+	if (qp->recvq)
+		munmap(qp->recvq, qp->num_rqe * sizeof(struct siw_rqe));
+
+	free(qp);
+
+	return NULL;
+}
+
+static int siw_modify_qp(struct ibv_qp *base_qp, struct ibv_qp_attr *attr,
+			 int attr_mask)
+{
+	struct ibv_modify_qp cmd;
+	struct siw_qp *qp = qp_base2siw(base_qp);
+	int rv;
+
+	memset(&cmd, 0, sizeof(cmd));
+
+	pthread_spin_lock(&qp->sq_lock);
+	pthread_spin_lock(&qp->rq_lock);
+
+	rv = ibv_cmd_modify_qp(base_qp, attr, attr_mask, &cmd, sizeof(cmd));
+
+	pthread_spin_unlock(&qp->rq_lock);
+	pthread_spin_unlock(&qp->sq_lock);
+
+	return rv;
+}
+
+static int siw_destroy_qp(struct ibv_qp *base_qp)
+{
+	struct siw_qp *qp = qp_base2siw(base_qp);
+	int rv;
+
+	pthread_spin_lock(&qp->sq_lock);
+	pthread_spin_lock(&qp->rq_lock);
+
+	if (qp->sendq)
+		munmap(qp->sendq, qp->num_sqe * sizeof(struct siw_sqe));
+	if (qp->recvq)
+		munmap(qp->recvq, qp->num_rqe * sizeof(struct siw_rqe));
+
+	rv = ibv_cmd_destroy_qp(base_qp);
+	if (rv) {
+		pthread_spin_unlock(&qp->rq_lock);
+		pthread_spin_unlock(&qp->sq_lock);
+		return rv;
+	}
+	pthread_spin_unlock(&qp->rq_lock);
+	pthread_spin_unlock(&qp->sq_lock);
+
+	free(qp);
+
+	return 0;
+}
+
+static struct ibv_ah *siw_create_ah(struct ibv_pd *pd, struct ibv_ah_attr *attr)
+{
+	return NULL;
+}
+
+static int siw_destroy_ah(struct ibv_ah *ah)
+{
+	return -EOPNOTSUPP;
+}
+
+static void siw_async_event(struct ibv_context *ctx,
+			    struct ibv_async_event *event)
+{
+	struct ibv_qp *base_qp = event->element.qp;
+	struct ibv_cq *base_cq = event->element.cq;
+
+	switch (event->event_type) {
+	case IBV_EVENT_CQ_ERR:
+		printf("libsiw: CQ[%d] event: error\n",
+		       cq_base2siw(base_cq)->id);
+		break;
+
+	case IBV_EVENT_QP_FATAL:
+		printf("libsiw: QP[%d] event: fatal error\n",
+		       qp_base2siw(base_qp)->id);
+		break;
+
+	case IBV_EVENT_QP_REQ_ERR:
+		printf("libsiw: QP[%d] event: request error\n",
+		       qp_base2siw(base_qp)->id);
+		break;
+
+	case IBV_EVENT_QP_ACCESS_ERR:
+		printf("libsiw: QP[%d] event: access error\n",
+		       qp_base2siw(base_qp)->id);
+		break;
+
+	case IBV_EVENT_SQ_DRAINED:
+	case IBV_EVENT_COMM_EST:
+	case IBV_EVENT_QP_LAST_WQE_REACHED:
+		break;
+
+	default:
+		break;
+	}
+}
+
+#define _load_mmapped(a)	\
```

Comment:
why do we need these wrappers?

Reply:
Probably not anymore. will change it.
I did this a while back when atomics were not available on some systems.